[ZIP 312] Specify key generation, change randomizer handling (#895)
zcash/zips: zips/zip-0312.rst

Reviewer: ```suggestion
The key word "MUST" in this document is to be interpreted as described
in BCP 14 [#BCP14]_ when, and only when, it appears in all capitals.
```

Reviewer: ```suggestion
Let ``FROST.verify_signature_share`` be the regular FROST signature share
verification function with arguments
``(identifier, PK_i, comm_i, sig_share_i, commitment_list, group_public_key, msg)``,
as defined in [#frost-aggregate]_. The ``randomized_verify_signature_share``
function takes an additional input ``randomizer`` computed as above, and adjusts
the inputs relative to this randomizer:

::

  def randomized_verify_signature_share(identifier, PK_i, comm_i, sig_share_i,
                                        commitment_list, group_public_key, msg):
    group_offset = G.ScalarBaseMult(randomizer)
    randomized_PK_i = PK_i + group_offset
    randomized_group_public_key = group_public_key + group_offset
    return FROST.verify_signature_share(identifier, randomized_PK_i, comm_i, sig_share_i,
                                        commitment_list, randomized_group_public_key, msg)
```

  ZIP: 312
  Title: FROST for Spend Authorization Multisignatures
  Owners: Conrado Gouvea <conrado@zfnd.org>
  Original-Authors: Chelsea Komlo
                    Deirdre Connolly
  Status: Draft
  Category: Wallet

>   Created: 2024-08-dd

Reviewer: This can just be the date it was originally committed which was apparently 2023-01-12. https://github.com/zcash/zips/commit/48f8c2c80e890385a9619e4dbf3b62ff8963b010
```suggestion
Created: 2023-01-12
```

  ZIP: 312
  Title: FROST for Spend Authorization Multisignatures
  Owners: Conrado Gouvea <conrado@zfnd.org>
          Daira-Emma Hopwood <daira-emma@electriccoin.co>
  Original-Authors: Chelsea Komlo
                    Deirdre Connolly
  Status: Draft
  Category: Wallet
  Created: 2022-08-01
  License: MIT
  Discussions-To: <https://github.com/zcash/zips/issues/382>
  Pull-Request: <https://github.com/zcash/zips/pull/662>
                <https://github.com/zcash/zips/pull/895>

Terminology
===========

The key words "MUST", "SHOULD", "RECOMMENDED", "NOT RECOMMENDED", and "MAY" in
this document are to be interpreted as described in BCP 14 [#BCP14]_ when, and
only when, they appear in all capitals.

The term "Recovery Protocol" in this document is to be interpreted as described

> in [#draft-ecc-quantum-recoverability]_.

Reviewer 2: Update this to point to ZIP 2005 instead.

- The zero-knowledge proof for the transaction is created with the randomizer
  as an auxiliary (secret) input, among others.

> When employing re-randomized FROST as specified in this ZIP, the goal is to

Reviewer: Right, because we're describing the variant that *is* re-randomized.

- The Coordinator is trusted with the privacy of the transaction (which includes
  the unlinkability property). A rogue Coordinator will be able to break
  unlinkability and privacy, but should not be able to create signed
  transactions without the approval of ``MIN_PARTICIPANTS`` participants, as
  specified in FROST.
- All key share holders are also trusted with the privacy of the transaction,
  thus a rogue key share holder will be able to break its privacy and
  unlinkability.

Caveats concerning the security model for use of FROST keys in a Quantum
Recovery protocol are covered in

> [#draft-ecc-quantum-recoverability-usage-with-frost]_.

Reviewer 2: Update this reference.

the trusted dealer. This allows backing up the unsplit key, but any benefit of
being able to do so does not outweigh the disadvantages of requiring a trusted
dealer under normal circumstances.

Regardless of how the Orchard or Sapling key components are generated, the
participants MUST ensure that they have obtained the same values for
$\mathsf{ak}$, $\mathsf{nk}$, and $\mathsf{rivk\_ext}$.

Since Sapling will not be modified to support Quantum Recoverability
[#draft-ecc-quantum-recoverability]_, Orchard is RECOMMENDED for new FROST keys.

In order for all participants to agree on the value of $\mathsf{sk}$, one
of the following options SHOULD be carried out:

- One participant generates a random $\mathsf{sk}$ and sends it to the
  others via an encrypted and authenticated channel. This option SHOULD be used
  if trusted dealer key generation was used, and the $\mathsf{sk}$ value
  can be sent along with the FROST shares during key generation. This option MAY
  be used if Distributed Key Generation was used, if participants find it
  acceptable to trust one participant to correctly generate $\mathsf{sk}$.

> - Participants generate a $\mathsf{sk}$ in distributed manner using TODO.

Reviewer: This should also reference https://c2sp.org/cocktail-dkg .

  also be as described in [#frost-primeordergroup]_ (where "raising an error"
  corresponds to returning $\bot$).

Note that the notation for elliptic-curve scalar multiplication,
``G.ScalarMult(P, k)``, differs from that used in the Zcash Protocol
Specification, which would use $[k]\, P$ with the group implied by $P$
[#protocol-abstractgroup]_.

Key Generation
--------------

FROST key generation with a trusted dealer is described in [#frost-tdkg]_. That
description does not support key re-randomization but can be adapted to do so,
as described in section 5 of [#frost-rerandomized]_.

However, trusted dealer key generation has the severe disadvantage that it
requires absolutely trustworthy hardware and software to be used by the dealer
in order to reliably avoid exposing the private key material. That is, it is not
sufficient simply to have an individual or organization that is trusted by all
participants; their entire hardware and software stack also needs to be trusted
by all participants. Since this will often be an unrealistic assumption that
partly defeats the point of using a threshold signature scheme, trusted dealer

> key generation SHOULD NOT be used.

Reviewer 3: ```suggestion
Trusted dealer key generation has the severe disadvantage that it
requires absolutely trustworthy hardware and software to be used by the dealer
in order to reliably avoid exposing the private key material. That is, it is not
sufficient simply to have an individual or organization that is trusted by all
participants; their entire hardware and software stack also needs to be trusted
by all participants. Since this will often be an unrealistic assumption that
partly defeats the point of using a threshold signature scheme, trusted dealer
key generation SHOULD NOT be used.
```

And make this a WARNING section

Instead, it is RECOMMENDED to instead use Distributed Key Generation. This is
supported by the Zcash Foundation's reference implementation of FROST
[#zf-frost-dkg]_.

To define a spending or viewing key that uses FROST, the Orchard and Sapling key
trees [#protocol-orchardkeycomponents]_ [#protocol-saplingkeycomponents]_ are
adjusted as follows:

- The Spend validating key $\mathsf{ak}$ is replaced by the FROST group
  public key `PK` [#frost-protocol]_ when deriving other key components.
- The Spend authorizing key $\mathsf{ask}$ [#protocol-spendauthsig]_ is
  replaced by the logical signing key that corresponds to the group public key
  `PK`. By design, this key is never explicitly constructed (unless trusted
  dealer key generation was used), and instead is represented by any sufficient
  subset of the participant FROST signing key shares `sk_i`.

The remaining parts of the Orchard and Sapling key trees SHOULD be generated
from a common spending key $\mathsf{sk}$ as described in
[#protocol-orchardkeycomponents]_ [#protocol-saplingkeycomponents]_, with the
exception of $\mathsf{ask}$ which SHOULD NOT be derived from $\mathsf{sk}$ (and

> can't possibly be derived, if Distributed Key Generation was used).

Reviewer 3: Remove the redundancy between here and the paragraph starting at line 201.

‘Orchard Key Components’ [#protocol-orchardkeycomponents]_ as modified by
[#draft-ecc-quantum-recoverability-usage-with-frost]_.

Similarly, when generating a FROST key for the Sapling shielded protocol, the
remaining parts of the Sapling key tree apart from $\mathsf{ask}$ and
$\mathsf{ak}$ (that is, $\mathsf{nk}$, $\mathsf{rivk\_ext}$, and further values
derived from them and $\mathsf{ak}$) SHOULD be generated from a common value
$\mathsf{sk}$ (again, not sufficient by itself for spend authorization) as
described in [#protocol-saplingkeycomponents]_.

If trusted dealer key generation is used contrary to recommendation, then
$\mathsf{ask}$ MAY in that case be derived from an $\mathsf{sk}$ known only to
the trusted dealer. This allows backing up the unsplit key, but any benefit of
being able to do so does not outweigh the disadvantages of requiring a trusted
dealer under normal circumstances.

Regardless of how the Orchard or Sapling key components are generated, the
participants MUST ensure that they have obtained the same values for
$\mathsf{ak}$, $\mathsf{nk}$, and $\mathsf{rivk\_ext}$.

Since Sapling will not be modified to support Quantum Recoverability

> [#draft-ecc-quantum-recoverability]_, Orchard is RECOMMENDED for new FROST keys.

Reviewer 3: Note that `#draft-ecc-quantum-recoverability` provides the rationale for this choice.

    absolutely trustworthy hardware and software to be used by the dealer in
    order to reliably avoid exposing the private key material. That is, it is
    not sufficient simply to have an individual or organization that is trusted
    by all participants; their entire hardware and software stack also needs to
    be trusted by all participants.

    Since this will often be an unrealistic assumption that partly defeats the
    point of using a threshold signature scheme, trusted dealer key generation
    SHOULD NOT be used.

Instead, it is RECOMMENDED to instead use Distributed Key Generation.

A suitable DKG protocol is COCKTAIL-DKG [#cocktail-dkg]_, which is supported by
the Zcash Foundation's reference implementation of FROST [#zf-frost-dkg]_.
COCKTAIL-DKG assumes that each participant has a static public key known to all
other participants. As noted in its documentation, the reference implementation
also requires participants to broadcast the commitment values honestly (i.e.
participants do not provide different commitment values to a subset of
participants), over a secure broadcast channel [#frost-term-broadcast-channel]_.

> Other Distributed Key Generation protocols MAY be used.

Reviewer 2: @conradoplg Does the ZF reference impl actually implement ChillDKG (on which COCKTAIL-DKG is based), or does it implement the DKG from the original FROST paper? If the latter, is that actually compatible with ChillDKG?

Without confirmation to the affirmative (especially because COCKTAIL-DKG is not finished), we should not claim it:
```suggestion
A suitable DKG protocol is COCKTAIL-DKG [#cocktail-dkg]_.
COCKTAIL-DKG assumes that each participant has a static public key known to all
other participants.

Other Distributed Key Generation protocols MAY be used. For example, the Zcash
Foundation's reference implementation of FROST [#zf-frost-dkg]_ includes a DKG
implementation for some ciphersuites. As noted in its documentation, the reference
implementation requires participants to broadcast the commitment values honestly
(i.e. participants do not provide different commitment values to a subset of
participants), over a secure broadcast channel [#frost-term-broadcast-channel]_.
```

EDIT: https://github.com/ZcashFoundation/frost/blob/main/frost-core/src/keys/dkg.rs makes it clear that it implements the DKG from the original FROST paper. So we need confirmation about whether or not that is exactly compatible with COCKTAIL-DKG.

Author: Apologies, I missed that. No we don't support COCKTAIL-DKG (though you could build it on top of the libraries)

dealer under normal circumstances.

Regardless of how the Orchard or Sapling key components are generated, the
participants MUST ensure that they have obtained the same values for
$\mathsf{ak}$, $\mathsf{nk}$, and $\mathsf{rivk\_ext}$.

Since Sapling will not be modified to support Quantum Recoverability
[#zip-2005]_, Orchard is RECOMMENDED for new FROST keys.

In order for all participants to agree on the value of $\mathsf{sk}$, one
of the following options SHOULD be used:

- One participant generates a random $\mathsf{sk}$ and sends it to the
  others via an encrypted and authenticated channel. This option SHOULD be used
  if trusted dealer key generation was used, and the $\mathsf{sk}$ value
  can be sent along with the FROST shares during key generation. This option MAY
  be used if Distributed Key Generation was employed, if participants find it
  acceptable to trust one participant to correctly generate $\mathsf{sk}$.
- Participants generate a $\mathsf{sk}$ in a contributory manner, for example
  using the method defined in `Contributory Generation of sk`_. This approach
  SHOULD be used to generate $\mathsf{sk}$ if Distributed Key Generation was

>   employed.

Author: Done in 31661fb00

Contributory Randomness Generation
----------------------------------

The following procedure allows generating randomness with inputs from all
participants, such that each participant can verify that its contribution
was included and that other honest participants will obtain the same value.

This is used to generate FROST randomizers, and can also be used for the
initial generation of $\mathsf{sk}$.

Two functions are provided for this purpose: $\mathtt{randomness\_generate}()$
and $\mathtt{randomness\_regenerate}()$. Both use helper functions defined as
follows:

- The $\mathtt{encode\_contributions}()$ function returns a byte serialization
  of a contribution list as defined below.
- $\mathtt{random\_bytes}(n)$ is defined in [#frost-conventions]_. It returns a
  sequence of $n$ bytes sampled uniformly at random.

A contribution list is a value of the form $[(i, \mathtt{contribution}_{\,i}),\, \ldots]$.
Each element in the list indicates a NonZeroScalar identifier $i$ and a byte

> sequence $\mathtt{contribution}_{\,i}$. This list MUST be sorted in ascending

Reviewer 2: Clarify that all `contribution_i`s in a contribution list are the same length (so the encoding is well-defined).

Author: Now moot

    ciphersuite).

  Outputs: randomness, a Scalar

  def randomness_regenerate(randomness_seed, contribution_list, H'):
    randomness_input = randomness_seed || encode_contribution_list(contribution_list)
    return H'(randomness_input)

Contributory Generation of sk
-----------------------------

This section describes a method of generating $\mathsf{sk}$ based on randomness
contributed by each participant, as well as the Coordinator. This avoids
entirely depending on the Coordinator's random number generator for security of
this key. This method depends on an encrypted, authenticated channel between the
Coordinator and each participant.

We make use of the following ciphersuite-generic parameters defined in the
`Ciphersuites`_ section:

- $\mathtt{Nh}$, the output size of the ciphersuite hash $\mathtt{H}$.
- $\mathtt{Hs}(m)$, a hash function.

Define::

  def sk_generate(contribution_list):
    return randomness_generate(contribution_list, Nh, Hs)

  def sk_regenerate(sk_seed, contribution_list):
    return randomness_regenerate(sk_seed, contribution_list, Hs)

The parties agree on $\mathsf{sk}$ as follows:

1. The Coordinator chooses a random seed

>    $\mathtt{sk\_seed} = \mathtt{random\_bytes}(\mathtt{Nh})$.

Reviewer 2: Bug: this `sk_seed` is not used.
```suggestion
```

2. Each participant $i$ generates
   $\mathtt{contribution}_{\,i} = \mathtt{random\_bytes}(\mathtt{Nh})$, and sends
   its contribution to the Coordinator over an encrypted, authenticated channel.
3. The Coordinator calls $\mathtt{sk\_generate}(\mathtt{contribution\_list})$,
   and sends the resulting $\mathtt{sk\_seed}$ to each participant over an
   encrypted, authenticated channel.
4. All parties regenerate $\mathsf{sk}$ using
   $\mathtt{sk\_regenerate}(\mathtt{sk\_seed}, \mathtt{contribution\_list})$,

>    and they include a Pedersen commitment to it in the transcript.

Reviewer 2: Bug: the other participants cannot run this, as they do not learn the `contribution_i` values from the other participants due to (in the current spec) them not being broadcast.

Author: Done in be04583, though we need to discuss the exact mechanism

  public key `PK` [#frost-protocol]_ when deriving other key components.
- The Spend authorizing key $\mathsf{ask}$ [#protocol-spendauthsig]_ is
  replaced by the logical signing key that corresponds to the group public key
  `PK`. By design, this key is never explicitly constructed (unless trusted
  dealer key generation was used), and instead is represented by any sufficient
  subset of the participant FROST signing key shares `sk_i`.

For both Orchard and Sapling, use of Distributed Key Generation precludes
generating $\mathsf{ask}$ from a spending key $\mathsf{sk}$. Instead the
Distributed Key Generation protocol will output the FROST group public key `PK`
(equivalent to $\mathsf{ak}$).

When generating FROST key components for the Orchard shielded protocol, the
remaining parts of the Orchard key tree apart from $\mathsf{ask}$ and
$\mathsf{ak}$ SHOULD be generated as described in [#zip-2005-usage-with-frost]_.
The specification there is normative, and is only summarized here. It requires
the participants to privately agree on a value $\mathsf{sk}$ (which is no longer
sufficient by itself for spend authorization), and then use it to obtain the

> nullifier deriving key $\mathsf{nk}$, the quantum spending key $\mathsf{qsk}$,

Reviewer 4: Isn't `sk` sufficient for spend authorization in the QR recovery case though? 

So in the recovery protocol, the threshold security assumption drops?

Reviewer 4: Perhaps its `sk` + `quantum-computer access`, so with a QC, any one multisig member could steal funds.

If this understanding is correct, we should explicitly note this. I wonder if its worth suggesting to use generic MPC tools like SPDZ w/ PQ-OT. (I think this should not block merge of the ZIP, and just be a note. We can ask folks like stoffelMPC)

Reviewer 4: cref: https://x.com/badcryptobitch/status/2055777683343130849?s=20

Reviewer: ZIP 2005 (https://zips.z.cash/zip-2005#usagewithfrost) already says that.

> Isn't sk sufficient for spend authorization in the QR recovery case though?

No, because you also need to sign for the FROST group key. If you have a CRQC then you can do that (but that is precisely what ZIP 2005 says).

Algorithms in this section are specified using Python pseudo-code, in the same
fashion as the FROST specification [#FROST]_.

> 

Reviewer 2: Remove the corresponding instantiations of `HR` below.

desirable in the context of FROST).

Randomizer Generation
---------------------

Re-Randomized FROST uses randomizers. This section specifies how they are
generated; this will be required for the Signature Share Generation
specification below.

Two functions are provided to generate randomizers: `randomizer_generate()` and
`randomizer_regenerate()`. Both use helper functions and a constant which are
defined as follows:

- The `encode_group_commitment_list()` function is defined in
  [#frost-listoperations]_. It returns a byte serialization of a
  `commitment_list` value.
- The `random_bytes(n)` function is defined in [#frost-conventions]_ and it
  returns a buffer with `n` bytes sampled uniformly at random.
- The `Ns` constant is define in [#frost-primeordergroup]_ and is the size of a
  serialized scalar.
- The `H2(m)` function is a ciphersuite-generic function defined in

>   [#frost-hash]_ but it is instantiated in the Ciphersuites section.

Reviewer: ```suggestion
  [#frost-hash]_ but it is instantiated in the `Ciphersuites`_ section.
```

::

  randomizer_generate():

  Inputs:
  - commitment_list = [(i, hiding_nonce_commitment_i,
    binding_nonce_commitment_i), ...], a list of commitments issued by
    each participant, where each element in the list indicates a
    NonZeroScalar identifier i and two commitment Element values
    (hiding_nonce_commitment_i, binding_nonce_commitment_i). This list
    MUST be sorted in ascending order by identifier.

>   Outputs: (randomizer_seed, randomizer), a byte buffer and a Scalar

Reviewer: ```suggestion
  Outputs: (randomizer_seed, randomizer), a byte sequence and a Scalar
```

  def randomizer_generate(commitment_list):

>     # Generate a random byte buffer with the size of a serialized scalar

Reviewer: ```suggestion
    # Generate a random byte sequence with the size of a serialized scalar.
```

desirable in the context of FROST).

Re-Randomized FROST
-------------------

To add re-randomization to FROST, follow the specification [#FROST]_ with the
following modifications.

Round One - Commitment
''''''''''''''''''''''

> 

Reviewer: Typo just below here: "Roune One" -> "Round One"

Author: thanks, solved in 62833e6c

Randomizer Generation

> '''''''''''''''''''''

Reviewer 2: Move this above `Re-Randomized FROST` and make it a same-level section. So this ZIP's Specification section then has:
- Key Generation
- Randomizer Generation
- Re-Randomized FROST
  - Round One - Commitment
  - Round Two - Signature Share Generation
  - Signature Share Verification and Aggregation
- Ciphersuites
  - FROST(Jubjub, BLAKE2b-512)
  - FROST(Pallas, BLAKE2b-512)

which then means that the "Re-Randomized FROST" sub-section is just focused on modifications to FROST, and everything else is stuff canonically specified in ZIP 312 (although maybe parts of "Key Generation" move elsewhere to e.g. ZIP 32, but we'll figure that out in #883).

Author: Good point, done in c74bed50

Two new helper functions are defined, which generate a randomizer; the first is
used by the Coordinator and the second is used by participants. The

> `encode_signing_commitments` function SHOULD return the byte serialization of

Reviewer 2: This is somewhat confusing, given that `random_bytes` gets its own paragraph. Separate `encode_signing_commitments` into its own paragraph, and then maybe turn those two paragraphs into bullet points? or something that clarifies these are _references_ to functions defined elsewhere that are used by the two new helper functions, not the two new helper functions being defined. Maybe also import the references first, and then say "Two new helper functions are defined" immediately before the actual definitions.

Author: Agreed, I restructured this in c74bed50

the `commitment_list` value as the serialization of the map between identifiers
and signing commitments, as described in [#frost-serialization]_.
Implementations MAY choose another encoding as long as all values (the
identifier, binding nonce and hiding nonce for each participant) are

> unambiguously encoded.

Reviewer 2: There is a risk here that a Coordinator and Participant (or multiple Participants) might have different implementations, and attempt to use them in the same signing session. That would be incompatible and cause a failure, which violates the defintion of MAY in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt):
> 5. MAY   This word, or the adjective "OPTIONAL", mean that an item is
   truly optional.  One vendor may choose to include the item because a
   particular marketplace requires it or because the vendor feels that
   it enhances the product while another vendor may omit the same item.
   An implementation which does not include a particular option MUST be
   prepared to interoperate with another implementation which does
   include the option, though perhaps with reduced functionality. In the
   same vein an implementation which does include a particular option
   MUST be prepared to interoperate with another implementation which
   does not include the option (except, of course, for the feature the
   option provides.)

So we will need to specify the full encoding here.

Author: Done in c74bed50 by using the serialization function from the RFC

Round Two - Signature Share Generation
''''''''''''''''''''''''''''''''''''''

In Round Two, the Coordinator picks a message and sends it to each signer along
with the commitments received in Round One, over an authenticated and encrypted
channel. In Re-Randomized FROST, an additional value called `randomizer` needs
to be generated by the Coordinator and sent along with the message and

> commitments.

Reviewer 2: Does the math not work out if each participant generates a share of alpha and commits to it in round 1, then uses their share in round 2? If it didn't then I guess the way to remove the trust assumption on the Coordinator here would be to still do that commitment in round 1, and then have a round 1.5 that is just participants revealing their alpha shares, but that would turn this into a three-round protocol, or two-round with preprocessing (where participants precompute and save their shares of both R and alpha), which isn't ideal for some usage scenarios where the Coordinator _is_ trusted. Maybe that should be left as an option?

Reviewer: One way to reduce the requirements on the Coordinator is to have it generate the randomizer by hashing the message, commitments, and an extra seed. The participants then recalculate it using that hash. The Coordinator can grind on the seed in order to create a subliminal channel, but this means you're not trusting the Coordinator's RNG. (Also, no participant can grind on their contribution in order to influence the randomizer, because the Coordinator only reveals the seed after it has the message and commitments.)

Author: I tried @daira 's approach in 62833e6c

SIGHASH themselves from that data. However, the specific mechanism for that process

> is outside the scope of this ZIP.

Reviewer 2: I think the mechanism for using FROST for transaction spend auth should _not_ be outside the scope of this ZIP, given that this ZIP is "FROST for Spend Authorization Multisignatures" (rather than the smaller-scoped "Re-randomized FROST"). However, I agree that most of the content that mechanism would entail is what would go into creating and checking PCZTs, so what we should do here is incorporate by reference the ZIP that eventually contains that specification (https://github.com/zcash/zips/issues/693), and then either use it as appropriate directly in this section, or divide the specification into "Re-randomized FROST" (for more general consumption) and "Using Re-randomized FROST for Zcash transactions" (which then augments it further with the Zcash-specific logic like PCZT usage).

Author: I agree, let's leave this pending until we get more concrete on PCZTs

Reviewer: Oh it's even worse. That comment is *only visible on the "new experience" view*.

In Zcash, the `randomizer` is called :math:`\mathsf{alpha}` and MUST be
generated accordingly to the Zcash specification [#protocol-spendauthsig]_. Note
that the choice of :math:`\mathsf{alpha}` influences the SIGHASH computation, so
it is impossible to compute the `randomizer` based on the message (SIGHASH), as
suggested in [#frost-rerandomized]_. This is not an issue as long the
`randomizer` is generated uniformly at random, which is done in the Zcash

> specification.

Reviewer: It's an overspecification to imply that the randomizer needs to be generated uniformly at random. As suggested above, it could be chosen depending on the commitments sent in round 1.

Also, the spec actually says to generate the randomizer like this:

> Define $\mathsf{RedDSA.GenRandom} ⦂ () \rightarrow \mathsf{RedDSA.Random}$ as:
> * Choose a byte sequence $T$ uniformly at random on $\mathbb{B}^{\mathbb{Y}[(\ell_{\mathsf{H}} + 128) / 8]}$.
> * Return $\mathsf{H}^⊛(T)$.

To refine my suggestion, $T$ would be the "seed" and then the randomizer would be computed by all participants as $\mathsf{H}^⊛(T \hspace{0.2em}\mathsf{||}\hspace{0.2em} \mathsf{commitments}).$

Author: Great idea. I went ahead with it in 62833e6c

  Outputs: randomizer, a Scalar

  def randomizer_regenerate(randomizer_seed, commitment_list):
    signing_commitments_enc = encode_group_commitment_list(commitment_list)
    randomizer_input = randomizer_seed || signing_commitments_enc
    return H2(randomizer_input)

Re-Randomized FROST
-------------------

To add re-randomization to FROST, follow the specification [#FROST]_ with the

> following modifications.

Reviewer: ```suggestion
To add re-randomization to FROST, the signers MUST follow the specification
[#FROST]_ with the following modifications.
```

SIGHASH themselves from that data. However, the specific mechanism for that process
is outside the scope of this ZIP.

The randomized `sign()` function is defined as the regular FROST `sign()` function,
but its inputs modified relative to the `randomizer` (which is computed with
`randomizer_regenerate()`).

- ``sk_i = sk_i + randomizer``

> - ``group_public_key = group_public_key + G.ScalarBaseMult(randomizer)``

Reviewer: ```suggestion
Let `FROST.sign` be the regular FROST signing function with arguments
`(identifier, sk_i, group_public_key, nonce_i, msg, commitment_list)`,
as defined in [#frost-sign]_. The `randomized_sign` function takes an
additional input `randomizer_seed`, and adjusts the inputs relative to
the `randomizer` computed by `randomizer_regenerate()`:

::

  def randomized_sign(identifier, sk_i, group_public_key, nonce_i, msg,
                      commitment_list, randomizer_seed):
    randomizer = randomizer_regenerate(randomizer_seed, commitment_list)
    randomized_sk_i = sk_i + randomizer
    randomized_group_public_key = group_public_key + G.ScalarBaseMult(randomizer)
    return FROST.sign(identifier, randomized_sk_i, randomized_group_public_key,
                      nonce_i, msg, commitment_list)
```

Reviewer: This should use double-backticks. (I'd fix it but my browser is broken.)

as defined also in [#frost-aggregate]_.

The $\textsf{Rerandomized-FROST.verify\_signature\_share}$ function takes an
additional input $\mathtt{randomizer}$ computed as above, and adjusts the inputs
relative to this randomizer::

  def Rerandomized-FROST.verify_signature_share(identifier, PK_i, comm_i, sig_share_i,
                                                commitment_list, group_public_key, msg):
    group_offset = G.ScalarBaseMult(randomizer)
    randomized_PK_i = PK_i + group_offset
    randomized_group_public_key = group_public_key + group_offset
    return FROST.verify_signature_share(identifier, randomized_PK_i, comm_i, sig_share_i,
                                        commitment_list, randomized_group_public_key, msg)

Contributory Generation of sk
-----------------------------

The contributory generation of $\mathsf{sk}$ uses the application-defined
payloads feature of the COCKTAIL-DKG protocol [#cocktail-dkg]_, as follows:

- In Round 1, generate ``sk_i = random_bytes(32)``
- Also in Round 1, set $\mathsf{payload}_{i,j} = \mathtt{sk\_i}$ for all `j`
- When running Round 3, use the following as the ``extension`` argument:

>   $ext = H(n \parallel len(payload_1) \parallel payload_1 \parallel \cdots \parallel len(payload_n) \parallel payload_n)$

Reviewer 2: Check whether this was meant to refer to a COCKTAIL-DKG payload (i.e. committing to other parts of the transcript) or `sk_i` (i.e. just the randomness from each participant). If the latter, then each `sk_i` could potentially be HD-derived instead of random, although there is already additional information that each signer needs to store beside their wallet seed phrase in order to use a threshold account.

Reviewer 2: From a call today with @conradoplg and nikkolasg: If we do a pure hash of concatenated `sk_i` then the last participant can bias `sk` (though to actually benefit from it, they would need to grind to bias `nk` or `qsk` via their chained hashes, as `sk` isn't used for anything directly). A way to avoid this issue could be to take the COCKTAIL-DKG transcript as the final input (like the random beacon we used in the Sapling proving key DKG).

Relatedly, nikkolasg had asked why not just derive `sk` from `ak`. I pointed out that this would break Sapling proof authorization keys, and it would mean that Orchard FVKs are fully derivable solely from `ak` instead of the FVK components being independent. (In current practice, `rivk` is derived from the other components due to ZIP 2005, but (`ak, nk, qk)` are still independent.)

Given that `ak` is sort-of-but-not-quite derived from the transcript, it might be fine to use `ak` as the final "random beacon" input instead of (a digest of) the transcript. It would still mean that `(nk, qk)` are in part derived from `ak`, but the fresh randomness from `sk_i` might mean they are sufficiently independent? I'm gonna think about it.

The security of Re-Randomized FROST with respect to the security assumptions of
regular FROST is shown in [#frost-rerandomized]_.

Regarding randomizer handling, in Zcash, the `randomizer` is called
:math:`\mathsf{alpha}` and is usually generated using the `RedDSA.GenRandom`

> function as defined inthe  Zcash specification [#protocol-spendauthsig]_. Note

Reviewer 2: ```suggestion
function as defined in the Zcash specification [#protocol-spendauthsig]_. Note
```

.. [#frost-tdkg] `RFC 9591: The Flexible Round-Optimized Schnorr Threshold (FROST) Protocol for Two-Round Schnorr Signatures. Appendix B: Trusted Dealer Key Generation <https://www.rfc-editor.org/rfc/rfc9591.html#name-trusted-dealer-key-generati>`_
.. [#frost-randomscalar] `RFC 9591: The Flexible Round-Optimized Schnorr Threshold (FROST) Protocol for Two-Round Schnorr Signatures. Appendix C: Random Scalar Generation <https://www.rfc-editor.org/rfc/rfc9591.html#name-random-scalar-generation>`_
.. [#frost-serialization] `The ZF FROST Book, Serialization Format <https://frost.zfnd.org/user/serialization.html>`_
.. [#frost-listoperations] `RFC 9591: The Flexible Round-Optimized Schnorr Threshold (FROST) Protocol for Two-Round Schnorr Signatures. Section 4.3: List Operations <https://www.rfc-editor.org/rfc/rfc9591.html#name-list-operations>`_
.. [#frost-conventions] `RFC 9591: The Flexible Round-Optimized Schnorr Threshold (FROST) Protocol for Two-Round Schnorr Signatures. Section 2: Conventions and Definitions <https://www.rfc-editor.org/rfc/rfc9591.html#name-conventions-and-definitions>`_

> .. [#frost-hash] `RFC 9591: The Flexible Round-Optimized Schnorr Threshold (FROST) Protocol for Two-Round Schnorr Signatures. Section 3.2: Cryptographic Hash Function <https://www.rfc-editor.org/rfc/rfc9591.html#name-cryptographic-hash-function>`_

Reviewer: ```suggestion
.. [#frost-hash] `RFC 9591: The Flexible Round-Optimized Schnorr Threshold (FROST) Protocol for Two-Round Schnorr Signatures. Section 3.2: Cryptographic Hash Function <https://www.rfc-editor.org/rfc/rfc9591.html#name-cryptographic-hash-function>`_
.. [#frost-sign] `RFC 9591: The Flexible Round-Optimized Schnorr Threshold (FROST) Protocol for Two-Round Schnorr Signatures. Section 5.2: Round Two - Signature Share Generation <https://www.rfc-editor.org/rfc/rfc9591.html#name-round-two-signature-share-g>`_
```
